library/std/src/thread/scoped.rs

```diff
@@ -0,0 +1,301 @@
+use super::{current, park, Builder, JoinInner, Result, Thread};
+use crate::fmt;
+use crate::io;
+use crate::marker::PhantomData;
+use crate::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
+use crate::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use crate::sync::Arc;
+
+/// A scope to spawn scoped threads in.
+///
+/// See [`scope`] for details.
+pub struct Scope<'env> {
+    data: ScopeData,
+    env: PhantomData<&'env ()>,
+}
+
+/// An owned permission to join on a scoped thread (block on its termination).
+///
+/// See [`Scope::spawn`] for details.
+pub struct ScopedJoinHandle<'scope, T>(JoinInner<'scope, T>);
+
+pub(super) struct ScopeData {
+    n_running_threads: AtomicUsize,
+    a_thread_panicked: AtomicBool,
+    main_thread: Thread,
+}
+
+impl ScopeData {
+    pub(super) fn increment_n_running_threads(&self) {
+        // We check for 'overflow' with usize::MAX / 2, to make sure there's no
+        // chance it overflows to 0, which would result in unsoundness.
+        if self.n_running_threads.fetch_add(1, Ordering::Relaxed) == usize::MAX / 2 {
```

Comment:
this should be `> usize::MAX / 2`, cuz with just equals we can still potentially invoke UB by continuously forgetting join handles fast enough that another increment happens between the `if` and the decrement/panic (thereby bypassing the check), then catching the panic and continuing on forgetting join handles
```suggestion
        if self.n_running_threads.fetch_add(1, Ordering::Relaxed) > usize::MAX / 2 {
```

Reply:
Thanks!